### PR TITLE
Microoptimization for UrlMatcher dumped as PHP code

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12099 |
| License | MIT |
| Doc PR | - |

I know that micro-optimizations are controversial. If you don't agree with it, feel free to close it.

---

The PHP code used to match URLs to routes repeats this structure a lot of times:

``` php
if (rtrim($pathinfo, '/') === '...') {
    if (substr($pathinfo, -1) !== '/') {
        return $this->redirect($pathinfo.'/', '...');
    }

    // ...
}
```

We could improve this code by introducing 3 variables that are calculated only once:

``` php
$isPathMissingTrailingSlash = substr($pathinfo, -1) !== '/';
$pathWithTrailingSlash = $isPathMissingTrailingSlash ? $pathinfo.'/' : $pathinfo;
$pathWithoutTrailingSlash = rtrim($pathinfo, '/');
```

Then the previous structure stops performing the same operations once and again:

``` php
if ($pathWithoutTrailingSlash === '...') {
    if ($isPathMissingTrailingSlash) {
        return $this->redirect($pathWithTrailingSlash, '...');
    }

    // ...
}
```

---

For other use cases this micro-optimization would be useless. But it could be worth for UrlMatcher because it is a critical piece of Symfony and it's executed for every request.

### CODE REVIEW

**Comment by jakzal** [comment]
Any chance to get some profiling results for this change?

**Comment by javiereguiluz** [comment]
@jakzal I'm on it.

**Comment by jvasseur** [comment]
I don't think precalculating `$isPathMissingTrailingSlash` and `$pathWithTrailingSlash` is useful because we always return after using them meaning they can't be used more than once.

**Comment by javiereguiluz** [comment]
@jvasseur you are right. I've removed those useless variables. Thanks!

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 1]
```diff
@@ -213,7 +214,7 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
 
         if (!count($compiledRoute->getPathVariables()) && false !== preg_match('#^(.)\^(?P<url>.*?)\$\1#', $compiledRoute->getRegex(), $m)) {
             if ($supportsTrailingSlash && substr($m['url'], -1) === '/') {
-                $conditions[] = sprintf("rtrim(\$pathinfo, '/') === %s", var_export(rtrim(str_replace('\\', '', $m['url']), '/'), true));
+                $conditions[] = sprintf("\$pathWithoutTrailingSlash === %s", var_export(rtrim(str_replace('\\', '', $m['url']), '/'), true));
```
Comment (stof) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: you should use a single-quoted string, avoiding the need to escape the ``$``

**Comment by fabpot** [comment]
@javiereguiluz Any evidence it helps with performance?

**Comment by javiereguiluz** [comment]
I'm closing this PR. I couldn't get Blackfire to show any significant improvement. My guess is that you need to have thousands of routes to make any difference.

I propose to close the related issue #12099 as "won't fix". Thanks!

**Pull request closed without merging**